liquidhaskell-boot/src/Language/Haskell/Liquid/Constraint/RewriteCase.hs

```diff
@@ -0,0 +1,125 @@
+        -- elaborated
+        go (EVar s1) e2 | isLocal s1 = Just [(s1, e2)]
+        go e1 (EVar s2) | isLocal s2 = Just [(s2, e1)]
+        -- TODO: Tecnically we could also unify under lambdas but you have to be
+        -- carefull about alpha equivalence idk if the effort is worth it.
+        go e1 e2 = do
+            -- Performing the unification under constructor is safe because 
+            -- C a₁ ... aₙ = C b₁ ... bₙ ⟺ ∀ n . a₁ = bₙ
+            EVar name1 : args1 <- pure $ flatten e1
+            EVar name2 : args2 <- pure $ flatten e2
+            guard $ name1 == name2
+            guard $ isCtor name1
+            guard $ length args1 == length args2
+            concat <$> zipWithM go args1 args2
+
+        isCtor  name = name `S.member` ctors
+        isLocal name = not (name `S.member` globals 
+                           || name `S.member` ctors
+                           || isPrefixOfSym anfPrefix name)
+
+flatten :: Expr -> [Expr]
+flatten (EApp f e) = flatten f ++ [e]
+flatten e = [e]
+
+groupUnifiableEqualities :: [(Expr, Expr)] -> [(Expr, Expr)]
+groupUnifiableEqualities = mapMaybe (keep2 . snd) . M.groupList 
+    where -- We perform only 2-way unification
+          keep2 [e1, e2] = Just (e1, e2)
+          keep2 _        = Nothing
+
+getEqualities :: Expr -> [(Expr, Expr)]
+getEqualities (PAtom Eq e1 e2) = [(e1, e2)]
+getEqualities (PAnd eqs)       = concatMap getEqualities eqs
+getEqualities _                = []
+
+
+-- +-----------------------------------------------------+
+-- | AcyclicRewrites: collection of rewrites that cannot |
+-- | cause an infinite loop                              |
+-- +-----------------------------------------------------+
+-- This could be extracted as a separate module
+
+-- | A collection of rewrites that cannot cause an infinite loop
+newtype AcyclicRewrites = AR (M.HashMap Symbol Expr)
+
+-- We can think of the map a directed graph where every symbol is a vertex and
+-- there is an edge v₁ → v₂ if in the expression that v₁ is rewritten to v₂ is a
+-- free variable.  To guarantee that there arent any infinite loops in the
+-- rewrite procedure that will be executed by PLE, we need to esnure that there
+-- aren't any cyclic rewrites, this is ensured by adding the invariant to our
+-- graph of beeing a DAG.
+
+-- | Takes a list of rewrites and returns a list of rewrites and filters out any
+-- rewrites that would cause an infinite loop, the procedure is order biased as
+-- rewrites earlier in the list take precedence.
+unloop :: [(Symbol, Expr)] -> LocalRewrites
+unloop = LocalRewrites . toRewrites . foldl' doInsert empty 
+    where doInsert ar (s, e) = ar `fromMaybe` insert ar s e 
+
+-- | Get the "raw" list of rewrites
+toRewrites :: AcyclicRewrites -> M.HashMap Symbol Expr
+toRewrites (AR m) = m
+
+-- | Checks if there is a path from s1 to s2 O(Σ(e ∈ m) |exprSymbolSet e|) [O(m)
+-- where m is the number of edges]
```

Comment:
```suggestion
-- | @existsPth g s1 s2@ yields @True@ checks if there is a path from @s1@ to @s2@
-- in @g@. Time is @O(Σ(e ∈ m) |exprSymbolSet e|)@, or said otherwise, it is @O(m)@
-- where @m@ is the number of edges.
```